### PR TITLE
Added feature for sending abandoned project mail

### LINKED ISSUE #3382: Surface "abandoned" state on projects in UI
We now have an inactive project policy: http://docs.readthedocs.io/en/latest/abandoned-projects.html

We should:
* Integrate a public facing "inactive" indicator into the RTD dashboard for projects that qualify

### PR DESCRIPTION
**Pull request by bansalnitish**
Fixes #3382 
@ericholscher I have added a button which is displayed when a project meets abandon policy and a user wants to use the name of the project. I am working on linking the button to send an email when it is clicked.

### CODE REVIEW

File: readthedocs/projects/forms.py [hunk 1]
```diff
@@ -114,7 +116,25 @@ def clean_name(self):
...
+            for user in project.users.all():
+                if user.is_superuser == True:
+                    email = user.email
+                    name = user
+            is_abandoned = False
+            has_good_build = project.has_good_build
+            if not has_good_build:
+               latest_build_date = project.get_latest_build().date
+               today = datetime.today()
+               diff = ((today.year - latest_build_date.year) * 12 +
```
Comment (stsewd) [inline readthedocs/projects/forms.py]: Is better if you use `timedelta` for date operations

File: readthedocs/projects/forms.py [hunk 2]
```diff
@@ -114,7 +116,25 @@ def clean_name(self):
         name = self.cleaned_data.get('name', '')
         if not self.instance.pk:
             potential_slug = slugify(name)
+            project = Project.objects.get(slug=potential_slug)
+            for user in project.users.all():
+                if user.is_superuser == True:
```
Comment (stsewd) [inline readthedocs/projects/forms.py]: This can be just

```python
if user.is_superuser:
    ...
```

**Comment by stsewd** [comment]
Can you add a screenshot of how this looks?

Also, I think it's better if you put a property to the `project` model, something like `is_abandoned`, so we can uses this on others part of the code.

**Comment by bansalnitish** [comment]
@stsewd Here is an screenshot:
image
Also, I will put the ``is_abandoned`` property to the ``project`` model itself!

**Comment by bansalnitish** [comment]
@ericholscher @stsewd I have added a feature for sending mail to the owner of abandoned projects when renaming is required. If a user enters a name which is already used and the project is abandoned, a button is displayed on top to send mail.

screenshot from 2018-03-03 21-57-29

If the user clicks on the button mail will be sent to the owner and user will be redirected to the import page with a ``Mail sent`` message.

screenshot from 2018-03-03 21-58-10

screenshot from 2018-03-03 21-57-40

File: readthedocs/projects/urls/private.py [hunk 3]
```diff
@@ -103,6 +103,10 @@
     url(r'^(?P<project_slug>[-\w]+)/advertising/$',
         ProjectAdvertisingUpdate.as_view(),
         name='projects_advertising'),
+
+    url(r'^import/manual/send_mail/$',
```
Comment (stsewd) [inline readthedocs/projects/urls/private.py]: Not sure if this should live on the project import page.

Comment (stsewd) [reply readthedocs/projects/urls/private.py]: But I do like this _hint_ when the project name is already taken

Comment (bansalnitish) [reply readthedocs/projects/urls/private.py]: Can you suggest something if we do not want this to stay on project import page ?

Comment (stsewd) [reply readthedocs/projects/urls/private.py]: Maybe showing this on the project page that is abandoned (only for logged users, of course). And a link to that project on the import page if the name is already taken by an abandoned project. Just a personal suggestion anyway, I'm don't know if the core maintainers want to go in that direction here.

Comment (stsewd) [reply readthedocs/projects/urls/private.py]: Also some tests for the `is_abandoned` property would be awesome.

Comment (bansalnitish) [reply readthedocs/projects/urls/private.py]: Hi @stsewd , Shall I make the changes for displaying the option of sending mail as suggested by you ?
Also can you please elaborate on the tests which will help reduce false results on the `is_abandoned` property.
Thanks

Comment (stsewd) [reply readthedocs/projects/urls/private.py]: From what Eric said

> I think we should keep it light weight in terms of implementation so this isn't a lot to maintain. Perhaps just an admin-only button that says "email project owner with our abandoned project email" as a start, which sends an email from a template that we have in the repo.

I think this button must be only shown to admins (rtd admins) and probably should be on the project page.

About the tests, you can create a project with different scenarios like failed build but not too old, too old but with successful build, etc. Probably the tests should live on this file https://github.com/rtfd/readthedocs.org/blob/master/readthedocs/rtd_tests/tests/test_project.py

Comment (bansalnitish) [reply readthedocs/projects/urls/private.py]: Okay Got it !
Thank you

**Comment by bansalnitish** [comment]
@stsewd How to run the tests written in `test_project.py` locally ? I was writing the tests as said by you but was not able to run them locally.

**Comment by stsewd** [comment]
@bansalnitish You can check https://docs.readthedocs.io/en/latest/tests.html, let me know if something isn't clear.

**Comment by bansalnitish** [comment]
Hi @stsewd, I have written the tests. Can you please have a look and suggest improvements!
Thanks :smile:

File: readthedocs/templates/core/project_bar_base.html [hunk 4]
```diff
@@ -23,6 +23,19 @@ <h1>
...
+    
+    {% if project.is_abandoned and request.user.is_superuser %}
+      <div class="abandon=project">
+        {% if not project.is_abandoned_mail_sent %}
+          <form action="{% url "send_mail" project.slug %}" method="post">
+            {% csrf_token %}
+            <input type="submit" name="submit-btn" value="{% trans "Email project owner with our abandoned project email" %}"/>
+          </form>
+        {% else %}
+          <p><strong>{%trans "Abandonment mail is sent to the owner of the project." %}</strong></p>
```
Comment (stsewd) [inline readthedocs/templates/core/project_bar_base.html]: _was sent_

File: readthedocs/projects/views/private.py [hunk 5]
```diff
@@ -264,6 +265,27 @@ def is_advanced(self):
...
 
 
+@login_required
+def send_mail(request, project_slug):
+    """Sends abandoned project email."""
+    project = Project.objects.get(slug=project_slug)
+    proj_name = project_slug
+    for user in project.users.all():
+        if AdminPermission.is_admin(user, project):
+            email = user.email
```
Comment (stsewd) [inline readthedocs/projects/views/private.py]: Can this be done on another way (not sure how we can _decide_ to which owner send the email)? Probably a `break` is needed here.

Comment (stsewd) [reply readthedocs/projects/views/private.py]: Also I think this is insecure, since any logged in user can send a request here, right?

Comment (bansalnitish) [reply readthedocs/projects/views/private.py]: I made this button to be visible only to a superuser in the html template. Does that look fine?

File: readthedocs/templates/core/project_bar_base.html (hunk 4 above)
Comment (bansalnitish) [reply readthedocs/templates/core/project_bar_base.html]: Done!

File: readthedocs/projects/views/private.py (hunk 5 above)
Comment (bansalnitish) [reply readthedocs/projects/views/private.py]: Also, I have added a ``break`` statement. Should the mail be sent to a list of all users?

Comment (stsewd) [reply readthedocs/projects/views/private.py]: Yes, but any user could do a manual request, so probably is worth to check for the superuser here. I'm not sure what is the best way of do it here anyway, probably the `required_login` decorator accepts a argument or there is a `super_user_only` decorator?

Maybe @humitos can help us here.

Comment (bansalnitish) [reply readthedocs/projects/views/private.py]: https://stackoverflow.com/questions/12003736/django-login-required-decorator-for-a-superuser
@stsewd Sorry for the delay in reply. Will this work for us?

Comment (stsewd) [reply readthedocs/projects/views/private.py]: That solution looks great! Let's wait to someone with more experience help us here, anyway, you could try testing this on your local instance and see if that works as expected.

Comment (bansalnitish) [reply readthedocs/projects/views/private.py]: Sure @stsewd! Thanks 😄

Comment (bansalnitish) [reply readthedocs/projects/views/private.py]: Sure @stsewd! Thanks 😄

Comment (bansalnitish) [reply readthedocs/projects/views/private.py]: @stsewd I tested this on my local instance and it works as expected!

Comment (bansalnitish) [reply readthedocs/projects/views/private.py]: @ericholscher Can you please guide me here!

File: readthedocs/projects/forms.py [hunk 6]
```diff
@@ -114,9 +114,16 @@ def clean_name(self):
...
-            if Project.objects.filter(slug=potential_slug).exists():
-                raise forms.ValidationError(
-                    _('Invalid project name, a project already exists with that name'))  # yapf: disable # noqa
+            project_exist = Project.objects.filter(slug=potential_slug).exists()
+            if project_exist:
+                project = Project.objects.get(slug=potential_slug)
+                project_url = project.get_absolute_url()
+                if project.is_abandoned:
+                    err_msg = ('Invalid project name, a <a href="{}" style="color: red">'
+                               'project</a> already exists with that name').format(project_url)
```
Comment (ericholscher) [inline readthedocs/projects/forms.py]: I don't think we need to check for abandoned, we should just *always* link to the project here.

File: readthedocs/projects/models.py [hunk 7]
```diff
@@ -558,6 +560,27 @@ def conf_dir(self, version=LATEST):
...
+                if diff > timedelta(days=365):
+                    return True
+                return False
+            return False
+        return True
+
+    @property
+    def is_abandoned_mail_sent(self):
+        """Is abandoned mail sent."""
+        return self.abandoned_mail_sent
```
Comment (ericholscher) [inline readthedocs/projects/models.py]: I don't believe this is necessary, you can just use `obj.abandoned_mail_sent` directly.

File: readthedocs/projects/urls/private.py [hunk 8]
```diff
@@ -36,6 +36,10 @@
         private.project_manage,
         name='projects_manage'),
 
+    url(r'^(?P<project_slug>[-\w]+)/send_mail/$',
+        private.send_mail,
+        name='send_mail'),
```
Comment (ericholscher) [inline readthedocs/projects/urls/private.py]: This should have a more explicit name & URL, `send_abandoned_mail` or similar

File: readthedocs/projects/views/private.py (hunk 5 above)
Comment (ericholscher) [reply readthedocs/projects/views/private.py]: This looks like it is checking for project owners who are admins. We just want to send the email to _all_ project owners, but only let *admins* do it. We should be doing a `request.user.is_superuser` or similar check to determine if the logged in user is an admin, and then send the email to all project owners.

Comment (bansalnitish) [reply readthedocs/projects/views/private.py]: The button for sending mail is only visible to the admins i.e. I have done a ``request.user.is_superuser`` in the html template. Also, we can check for superuser using a decorator like this: https://stackoverflow.com/questions/12003736/django-login-required-decorator-for-a-superuser. Does that look fine @ericholscher?

**Comment by bansalnitish** [comment]
@ericholscher I have made the requested changes. Can you have a look once! 
Thanks!

**Comment by agjohnson** [comment]
As per feedback in #3382, I'm closing this. The branch will still be available for reference if needed.

**Pull request closed without merging**